### DISCUSSION TITLE
How to detect if an update is available

### DISCUSSION THREAD

**Discussion by peternowell** [sparkle-project/Sparkle#2604] [Q&A] [answered]
I would like to add a small badge/button in my app's UI when a software update is available. When the user clicks on it, Sparkle's standard update UI (with release notes) should appear. I'm probably missing something very basic, but I'm not sure how my app should check whether an update is available without also making the Sparkle's UI appear. I'd love to simply have a `updateAvailable: Bool` variable on some class, that I can check from anywhere in my app and quietly refresh periodically. If it's helpful, I describe my basic Sparkle setup [here](https://github.com/sparkle-project/Sparkle/discussions/2559#discussioncomment-9449644). What do you recommend? Thanks!

**Comment by zorgiepoo** [comment, marked as answer]
https://sparkle-project.org/documentation/gentle-reminders/ is designed to handle this use case.

By default Sparkle handles scheduling update checks periodically (this is orthogonal to not showing the full UI actually). Gentle reminders is for overriding the presentation of the new update behavior when an update check is performed and a new update is available to be shown.

In the older days, some apps would do two update checks; one "silent background info" update check to see if a new update was available, then another update check when the user hits the badge. I don't recommend this approach because it's inefficient.

**Reply by peternowell** [reply]
Thank you! I just read through the docs and I think the 1st code example is very close to what I need. I was worried that gentle reminders were a replacement for showing Sparkle's standard UI but it seems set up to be complementary. I'll let you know if I have any further questions. 🙏🏼

**Comment by peternowell** [comment]
Ok, I've got the gentle reminders working, and it seems like its purpose is to make only the _scheduled_ update checks happen in a gentle/no-UI way. My implementation creates an `updateAvailable` property for my UI to use, and I set it to `true` when `standardUserDriverWillHandleShowingUpdate()` runs. But what I really wanted is a way to refresh/check for updates completely quietly/gently, at opportune times (ex: when a window or view loads that can show my update available indicator).

I discovered that I can do this by running the SPUUpdater's [`checkForUpdateInformation()`](https://sparkle-project.github.io/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdateInformation) method and implementing a custom SPUUpdaterDelegate that provides methods for [`-updater:didFindValidUpdate:`](https://sparkle-project.github.io/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:didFindValidUpdate:) and [`-updaterDidNotFindUpdate:error:`](https://sparkle-project.github.io/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updaterDidNotFindUpdate:error:) methods. With this solution, my app can run `checkForUpdateInformation()` whenever it seems like a good time.

I just wanted to check this solution with you, in case I misunderstood something. Thanks!

**Reply by zorgiepoo** [reply]
> I discovered that I can do this by running the SPUUpdater's [checkForUpdateInformation()](https://sparkle-project.github.io/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdateInformation) method and implementing a custom SPUUpdaterDelegate that provides methods for [-updater:didFindValidUpdate:](https://sparkle-project.github.io/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:didFindValidUpdate:) and [-updaterDidNotFindUpdate:error:](https://sparkle-project.github.io/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updaterDidNotFindUpdate:error:) methods. With this solution, my app can run checkForUpdateInformation() whenever it seems like a good time.

This is the "inefficient" way I was talking about, because whenever the user clicks the button Sparkle now has to perform another update check.

> But what I really wanted is a way to refresh/check for updates completely quietly/gently, at opportune times (ex: when a window or view loads that can show my update available indicator).

You can trigger a background scheduled check with [checkForUpdatesInBackground](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdatesInBackground). Note this is what Sparkle itself does automatically and periodically if you keep automatic checking of updates enabled.

Like I said the presentation of scheduled updates is orthogonal to performing update checks. You can use gentle reminders, and schedule a background update check when you want too.

**Reply by zorgiepoo** [reply]
That said, you may also want to consider if the user has automatic downloading/installation of updates (which happens on quit typically). checkForUpdatesInBackground may actually download/prepare installing the update rather than bring up a scheduled update notice (the first time) in that case. If you speed up the frequency of update checks, this could actually increase the UX burden (from silent updates --> updates that would have been silent but now have a badge indicator sooner than Sparkle's natural lifecycle).

checkForUpdateInformation and showing a badge on the other hand can sabotage/prevent automatically downloaded/installed updates, which is no good.
